Add syncthreads to scans prevent overwriting shared memory of previous chunk

ivogabe opened the pull request:
<!--
Hi!

Thanks for taking the time to create this pull request! You are awesome (:

The following schema may help when filing your pull request:
-->

## Description
<!--
Provide a description of the pull request here.
Try to include a general summary of the changes in the title above.
-->
This PR fixes an issue that we had for a long time. Programs with scans would sometimes give incorrect values. As I found out, this is caused by missing synchronization in thread groups in the scan kernels.

## Motivation and context
<!--
Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
-->
A thread group in a scan kernel may handle multiple chunks of the input. 
A __syncthreads was missing between the end of one chunk and the start of the next, which caused that
shared memory could be overwritten in in the next iteration, before all threads are finished reading data from the previous iteration.
scanBlockShfl does perform synchronisation between its writes and reads, but there was no synchronisation between the reads of some iteration of the outer loop, and the writes in the next iteration.

## How has this been tested?
<!--
Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc.
-->
Our quickhull implementation now works correctly on the GPU, whereas it previously would produce garbage values or crash (as it continues to compute with those garbage values). This was tested on an A100 and 4090.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--
Go over all the following points, and put an `x` in all the boxes that apply.
If you're unsure about any of these, don't hesitate to ask. We're here to help!
-->

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

tomsmeding commented:
With this PR, the following code infinite-loops on jizo:
```haskell
import qualified Data.Array.Accelerate as A
import qualified Data.Array.Accelerate.LLVM.PTX as GPU
main = print $ length $ show $ GPU.runN (A.scanl1 (+)) $ A.fromList (A.Z A.:. 257 A.:. 20) [1::Int ..]
```
This reproduces on jizo as long as the inner dimension is _not_ a multiple of 32 and the outer dimension is `>= 257`.
The problem is not specific to arrays of dimension 2; the thing that must be `>= 257` is the product of all non-inner dimensions. (Tested with arrays of dimension 3.)

The problem does not seem to occur on my 1050 Ti. On jizo, however, it occurs both with old+llvmhs and old+llvmpretty.

Let `N` be the size of the inner dimension (`20` above). Checking with `cuda-gdb` some time after the scan should long have been done, it turns out that:
- `N` threads are currently in a `WARPSYNC R7` instruction inside `__cuda_sm70_shflsync_up`, called from somewhere way down in the scan kernel;
- `32-N` threads are currently in a `WARPSYNC 0xffffffff` all the way at the start of the kernel, which (cross-referencing the SASS in `cuda-gdb` with the generated PTX, then the unoptimised LLVM IR, and finally the Haskell code in DAA.LLVM.PTX.CodeGen.Scan) seems to correspond to [this added `__syncthreads`](https://github.com/AccelerateHS/accelerate-llvm/pull/97/files#diff-e92694348032c17f09a6520de4a6b21a92a72f5fd6f390d7902a74b510b09919R194).

Perhaps you know where to go from here.

ivogabe commented:
With help from Tom I managed to finally locate the problem. There was already a problem with syncthreads before my changes. We already performed syncthreads under non-uniform control flow (which is not supported). However, we were somehow lucky and the kernel just ran. After my changes, which adds an extra syncthreads to the kernel, this existing bug became an actual problem.

After staring at the code for a long time (again with help from Tom) I saw that in a scanl1 (or scanr1), we would perform a syncthreads under non-uniform control flow if the innermost dimension is smaller than the threadblock size. These scans have no seed (initial value) and have a separate loop for the first elements (where we don't have an aggregate yet) and the remaining elements (where we do already have an aggregate). The latter properly made sure that we run scanBlock (which does the actual scanning in a block) under uniform control flow. However, the former did not.

This did make it a bit easier to fix this problem (after we finally found it), since the fix was already there, just a few lines below. The last commit on this PR fixes that. Furthermore, I added some comments to explain that the control flow is uniform at some key places, and I spotted another place where we did a syncthreads just after a non-uniform conditional (which I moved to before that conditional).

As far as I know, all known scan bugs should now be fixed.
